app/controllers/api/users.php

```diff
@@ -107,43 +110,60 @@
     ->label('sdk.response.code', Response::STATUS_CODE_OK)
     ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
     ->label('sdk.response.model', Response::MODEL_USER_LIST)
+    ->param('queries', [], new ArrayList(new Text(APP_LIMIT_ARRAY_ELEMENT_SIZE), APP_LIMIT_ARRAY_PARAMS_SIZE), 'Array of query strings generated using the Query class provided by the SDK. [Learn more about queries](https://appwrite.io/docs/databases#querying-documents). Maximum of ' . APP_LIMIT_ARRAY_PARAMS_SIZE . ' queries are allowed, each ' . APP_LIMIT_ARRAY_ELEMENT_SIZE . ' characters long. You may filter on the following attributes: ' . implode(', ', ListUsers::ALLOWED_ATTRIBUTES), true)
     ->param('search', '', new Text(256), 'Search term to filter your list results. Max length: 256 chars.', true)
-    ->param('limit', 25, new Range(0, 100), 'Maximum number of users to return in response. By default will return maximum 25 results. Maximum of 100 results allowed per request.', true)
-    ->param('offset', 0, new Range(0, APP_LIMIT_COUNT), 'Offset value. The default value is 0. Use this param to manage pagination. [learn more about pagination](https://appwrite.io/docs/pagination)', true)
-    ->param('cursor', '', new UID(), 'ID of the user used as the starting point for the query, excluding the user itself. Should be used for efficient pagination when working with large sets of data. [learn more about pagination](https://appwrite.io/docs/pagination)', true)
-    ->param('cursorDirection', Database::CURSOR_AFTER, new WhiteList([Database::CURSOR_AFTER, Database::CURSOR_BEFORE]), 'Direction of the cursor, can be either \'before\' or \'after\'.', true)
-    ->param('orderType', Database::ORDER_ASC, new WhiteList([Database::ORDER_ASC, Database::ORDER_DESC], true), 'Order result by ASC or DESC order.', true)
     ->inject('response')
     ->inject('dbForProject')
     ->inject('usage')
-    ->action(function (string $search, int $limit, int $offset, string $cursor, string $cursorDirection, string $orderType, Response $response, Database $dbForProject, Stats $usage) {
+    ->action(function (array $queries, string $search, Response $response, Database $dbForProject, Stats $usage) {
+
+        // Parse queries input and also extract cursor, if there is one
+        $cursor = null;
+        foreach ($queries as $i => $query) {
```

Comment:
I think we can abstract this logic with a helper method. Maybe `Utopia\Database\Query::getByType($queries, Query::TYPE_CURSORAFTER);`

Reply:
I was trying to avoid iterating over the array again.

Reply:
Sometime we have to prefer code maintenance over performance 🤷‍♂️

Reply:
In that case, what about:

```php
$grouped = Queries::groupByType($queries);
$cursor = $grouped['cursorDirection'] === Database::CURSOR_AFTER  ? Query::cursorAfter($grouped['cursor']) : Query::cursorBefore($grouped['cursor']);
```

to avoid creating another function?

Reply:
Good point. Let's think of a view variations and see if we can abstract some of this logic from the action.

Right now my main concerns:
1. This is a lot of repeatable code in the action which we'll need to implement in +10 routes?
2. Action complexity is higher, we always try to max their simplicity (not in every cost), for sake of readability and maintenance, performance will be less priority in many of this cases. Especially when impact is small. 
3. Separation of concerns is not followed, action is handling the input validation and not the param validators (we do have some rare cases where we do this, but mostly this is unusual). We also avoid this as future devs could easily start using not 100% validated variables, which is hard to spot in a review and very risky at times. (AKA "Developer proofing")
4. If we do abstract the logic to the DB library, the structure and names have to make sense and be complimentary to each other, not forced just in sake of a quick solution.

Let me know what you think.

Reply:
> 1. This is a lot of repeatable code in the action which we'll need to implement in +10 routes?

I definitely agree with trying to reduce duplicated code. I'm just not sure where to put things. Maybe as static methods in Appwrite's queries Validator?

> 2. Action complexity is higher, we always try to max their simplicity (not in every cost), for sake of readability and maintenance, performance will be less priority in many of this cases. Especially when impact is small.

Thanks for the clarification on the design philosophy 🙏 I'll try to keep this in mind

> 3. Separation of concerns is not followed, action is handling the input validation and not the param validators (we do have some rare cases where we do this, but mostly this is unusual). We also avoid this as future devs could easily start using not 100% validated variables, which is hard to spot in a review and very risky at times. (AKA "Developer proofing")

I can try to put the Validator in the params but I think we'll still need to parse the queries in the action body. Is that okay?

> 4. If we do abstract the logic to the DB library, the structure and names have to make sense and be complimentary to each other, not forced just in sake of a quick solution.

I'm not sure I like the idea of having the Database->find() method accept an array of strings. Do really like being able to pass an array of Query.